lib/Parse/Lexer.cpp

```diff
@@ -1621,6 +1623,38 @@ static size_t commonPrefixLength(StringRef shorter, StringRef longer) {
   return offset;
 }
 
+void Lexer::lexRegexLiteral(const char *TokStart) {
+  assert(*TokStart == '\'');
+
+  bool HadError = false;
+  while (true) {
+    // Check if we reached the end of the literal without terminating.
+    if (CurPtr >= BufferEnd || *CurPtr == '\n' || *CurPtr == '\r') {
+      diagnose(TokStart, diag::lex_unterminated_regex);
+      return formToken(tok::unknown, TokStart);
+    }
+
+    const auto *CharStart = CurPtr;
+    uint32_t CharValue = validateUTF8CharacterAndAdvance(CurPtr, BufferEnd);
+    if (CharValue == ~0U) {
+      diagnose(CharStart, diag::lex_invalid_utf8);
+      HadError = true;
+      continue;
+    }
+    if (CharValue == '\\' && (*CurPtr == '\'' || *CurPtr == '\\')) {
+      // Skip escaped delimiter or \.
+      CurPtr++;
+    } else if (CharValue == '\'') {
+      // End of literal, stop.
+      break;
+    }
+  }
+  if (HadError)
+    return formToken(tok::unknown, TokStart);
+
+  formToken(tok::regex_literal, TokStart);
+}
+
```

Comment:
(longer term) What do you think about pulling this functionality out of C++ and into the package? That is C++ passes in the character pointer and receives the terminating pointer for when it's done. That would allow us to mock up multiple kinds of literals, such as:

```
let regexen: Array<Regex> = [
  '/PCRE/',
  '|Modern|',
  '''
    multiline     // Line-ending comment
  ''',
  #'r'a'w'#
]
```  

```

Reply:
I think that would work, though it may be more complicated if we eventually decide to go with `/` delimiters, as that would likely require a bunch of lexing heuristics on the C++ side. But maybe that could be a special case, with other delimiters being passed straight through